**Universidad Galileo**

**Ciencia de Datos en Python**

**Nombre: Rodrigo Chang**

In [5]:
import validaciones

# Ejercicios Python

## Zen of Python

Lee una vez mas el Zen of Python antes de iniciar con la tarea

In [70]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## 1 - XOR

En la siguiente celda crea una funcion para la compuerta logica XOR 

<img src="assets/xor.png" width="400">


In [71]:
def xor(a,b):
    """
    Esta funcion recibe dos valores booleanos, ejemplo xor(1, 0) o True y False
    y retorna 0 o 1 como resultado de la operacion logica XOR
    """
    # tu codigo aqui
    return (a and (not b)) or ((not a) and b)

In [72]:
xor(0 , 1) 

1

In [73]:
xor(0 , 0)

0

In [74]:
xor(True , False)

True

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

In [75]:
validaciones.xor_function(xor)

Prueba 1
Prueba 1 [CORRECTA]
Prueba 2
Prueba 2 [CORRECTA]


## 2 - Sumando N series

Tenemos una secuencia cuyo $n^{th}$ es:
\begin{equation*}
\ T_n = n^2 - (n-1)^2
\end{equation*}

Hay que evaluar las series:
\begin{equation*}
\ S_n = T_1 + T_2 + ... + T_n
\end{equation*}


**Instrucciones**
En la siguiente celda en la funcion summingSeries debemos **retornar** en valor de $S_n$, tenemos como entrada n

**Ejemplo 1:**

Si la entrada es 2

Salida:
4

Explicacion:

$T_1 = 1$

$T_2 = 3$

$S_2 = T_1 + T_2$

$S_2 = 4 $



**Tip :**
Antes te implementar $S_n$ la solución, analicemos algebraicamente el problema para encontrar una solución equivalente optima.


## Solución

Resulta que si expandimos el binomio en $T_n$, tenemos que $$T_n = 2n-1$$

Luego, como $S_n = T_1 + T_2 + \ldots + T_n = \sum_{i=1}^{n} T_i = \sum_{i=1}^{n} (2i-1)$. Resolviendo la sumatoria, tenemos que 
$$ S_n = n^2 $$. 

In [76]:
def summatoriaSerie(n):
    """
    La funcion recibe como parametro n y retorna el valor se la serie
    """
    # tu codigo aqui
    # La función de sumatoria de S_n se puede reescribir como S_n = n*n
    return n**2

#### 2.1 - Ciclos

Crea un ciclo for que imprima el resultado de la funcion **summingSeries** desde 1 hasta 10

In [77]:
for i in range(1,11):
    print("Resultado de summingSeries hasta n = ", i, " = ", summatoriaSerie(i) )

Resultado de summingSeries hasta n =  1  =  1
Resultado de summingSeries hasta n =  2  =  4
Resultado de summingSeries hasta n =  3  =  9
Resultado de summingSeries hasta n =  4  =  16
Resultado de summingSeries hasta n =  5  =  25
Resultado de summingSeries hasta n =  6  =  36
Resultado de summingSeries hasta n =  7  =  49
Resultado de summingSeries hasta n =  8  =  64
Resultado de summingSeries hasta n =  9  =  81
Resultado de summingSeries hasta n =  10  =  100


### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [78]:
validaciones.summingSeries_function(summatoriaSerie)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]


## 3 - Factores Primos

Para cada $n$ queremos obtener el conteo maximo de numeros primos unicos en en rango inclusivo $[1, n]$,
y retornar el valor del conteo en una nueva linea

**Nota:** Recuerda que un numero primo solo es divisible por el mismo y que 1 no es un numero primo

##### Ejemplos:

###### Ejemplo 1
Entrada: 1

Salida esperada: 0

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,1]$ es $0$, porque $1$ no es un numero primo.

###### Ejemplo 2
Entrada: 3

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,3]$ es $1$, porque el numero $3$ tiene 1 factor primo unico (el mismo) 

###### Ejemplo 3
Entrada: 500

Salida esperada: 4

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,500]$ es $4$, porque el producto de los primeros cuatro numeros primos unicos es $2 \times 3 \times 5 \times 7 = 210$ y no hay ningun numero primo unico que multiplicado por el resultado sea $	\leqslant 500$



#### Tip: 
Utiliza la funcion `range()` de python


In [79]:
# Guardamos algunos números primos para ver cómo se comporta el producto del primer hasta el n-ésimo primo.
listaPrimos=[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199];

# Inicializamos el producto
res = listaPrimos[0];

# Vamos viendo cómo evoluciona el producto de los primeros 46 números primos
for i in range(1,len(listaPrimos)):
    res =  res * listaPrimos[i];
    print(i, res)

1 6
2 30
3 210
4 2310
5 30030
6 510510
7 9699690
8 223092870
9 6469693230
10 200560490130
11 7420738134810
12 304250263527210
13 13082761331670030
14 614889782588491410
15 32589158477190044730
16 1922760350154212639070
17 117288381359406970983270
18 7858321551080267055879090
19 557940830126698960967415390
20 40729680599249024150621323470
21 3217644767340672907899084554130
22 267064515689275851355624017992790
23 23768741896345550770650537601358310
24 2305567963945518424753102147331756070
25 232862364358497360900063316880507363070
26 23984823528925228172706521638692258396210
27 2566376117594999414479597815340071648394470
28 279734996817854936178276161872067809674997230
29 31610054640417607788145206291543662493274686990
30 4014476939333036189094441199026045136645885247730
31 525896479052627740771371797072411912900610967452630
32 72047817630210000485677936198920432067383702541010310
33 10014646650599190067509233131649940057366334653200433090
34 149218235093927932005887573661584106854758386

Como vemos, este producto es bastante grande, y podría servir (por ahora) para los propósitos de nuestra función.

In [80]:
def conteoPrimos(n):
    """
    Esta funcion recibe como parametro n y 
    retorna el conteo maximo de numeros primos unicos
    """
    # tu codigo aqui
    conteo = 0;
    res = listaPrimos[0];
    
    for i in range(1, len(listaPrimos)): 
        if (res > n):
            return conteo;
        
        res = res * listaPrimos[i];
        conteo += 1;

### Pruebas

In [81]:
conteoPrimos(1)

0

In [82]:
conteoPrimos(500)

4

### Otra versión

El problema con nuestra función anterior, es que depende de tener una lista de números primos, a continuación se presenta una implementación alternativa, en la cual esta lista es generada hasta donde se necesite para poder obtener los primos únicos.

In [1]:
def conteoPrimos(n):
    """
    Esta funcion recibe como parametro n y 
    retorna el conteo maximo de numeros primos únicos
    """
    # Definimos una función auxiliar 'esPrimo(n)', que devuelve True si n es primo.
    def esPrimo(n):
        for i in range(2,n):
            if n % i == 0:
                return False;
        return True;

    # Definimos algunos casos especiales
    if n == 1: 
        return 0;
    elif n <= 3:
        return 1;
    # Si el número es mayor a 3, entonces
    else:
        # Utilizamos una variable 'res' para llevar el producto
        res = 1;
        # 'cont' nos indica el número de primos que llevamos multiplicando
        cont = 0;

        # Para cada número 'elem' en la lista de 2 al número n:
        for i, elem in enumerate(range(2, n)):
            # Si 'elem' es primo
            if esPrimo(elem):
                # Lo multiplicamos por nuestro producto acumulado
                res *= elem;
                # Aumentamos la cuenta de que es primo
                cont += 1;
                # Si nuestro producto acumulado supera a n, terminamos
                if res > n:
                    # Devolvemos cont-1
                    return cont-1;
    

### Pruebas

In [2]:
conteoPrimos(1)

0

In [3]:
conteoPrimos(500)

4

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [6]:
validaciones.primeCount_function(conteoPrimos)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]
Prueba 7
Prueba 7[CORRECTA]
Prueba 8
Prueba 8[CORRECTA]
Prueba 9
Prueba 9[CORRECTA]
Prueba 10
Prueba 10[CORRECTA]
Prueba 11
Prueba 11[CORRECTA]
Prueba 12
Prueba 12[CORRECTA]


## 4 - Metodo de Euler

El metodo de Euler se utiliza para aproximar la solucion particular de una ecuacion diferencial dado un valor inicial. Con esta informacion se sabe que la grafica de esa solucion pasa a traves del punto $(x_0,y_0)$ y tiene una pendiente $F(x_0,y_0)$ en ese punto, esto da un punto de partida para aproximar la solucion.

A partir del punto inicial, se sigue en la direccion indicada por la pendiente. Mediante un pequeño paso $h$, se mueve a lo largo de la recta tangente hasta llegar al punto $(x_1,y_1)$ donde:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ y_1 = y_0 + hF(x_0,y_0)\\
\end{equation*}


<img src="assets/Euler.png" width="350">

Como se muestra en la figura se considera $(x_1,y_1)$ como un nuevo punto inicial.

Los valores de $x$ son:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ x_2 = x_1 + h       \\
\    .                \\
\    .                \\
\    .                \\
\ x_n = x_{n-1} + h   \\
\end{equation*}


Los valores de $y$ son:
\begin{equation*}
\ y_1 = y_0 + hF(x_0,y_0)\\
\ y_1 = y_1 + hF(x_1,y_1)\\
\       . \\
\       . \\
\       . \\
\ y_n = y_{n-1} + hF(x_{n-1},y_{n-1})\\
\end{equation*}


#### Problema 4
El ejecicio consiste en implementar la funcion *euler_metod* esta funcion recibe como parametro el 
* punto inicial $x_0$
* punto inicial $y_0$
* Paso $h$ 
* numero de iteraciones $n$ 
* La equacion diferencial que queremos aproximar, esta funcion tambien recibe los parametros $x$ y $y$
* La funcion de la solucion exacta que tambien tambien recibe los parametros $x$ y $y$

La funcion debe retornar el valor de $y_n$ y el error
$error=|y - y_n|$ 

In [2]:
import math 

In [6]:
def euler_method(x,y,h,n, diff_eq, func):
    """
    Parametros x, y punto inicial, logitud del paso h , n numero de iteraciones,
    diff_eq la equacion diferencial y func la funcion de la solucion exacta
    Retorna y y el error
    """
    # tu codigo aqui
    for i in range(1, n+1):
        # Actualizar los valores de x e y.
        x = x + h;
        y = y + h*diff_eq(x,y)
        
        # Calcular el error
        error = abs(func(x) - y) 
        
        # Mostrar un mensaje de estado
        print("x_%d = %0.4f, y_%d = %0.4f;  error = %0.4f" % (i, x, i, y, error))
        
    # Devuelve el valor final para y, además devuelve el error en la última iteración
    return y, error
   
        

### Ejemplo

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial y}{\partial x} = -2y \\
\end{equation*}
Donde $y(0) = 4$ vamos a unsa un de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ y = 4e^{-2x} \\
\end{equation*}

In [4]:
### Implementando las funciones del ejemplo
def y_prime(x, y):
    return -2*y

def y(x):
    return 4 * math.exp(-2*x)

In [5]:
"""
Llamamos a la funcion de Euler
euler_method(0,4,0.1,10, y_prime,y)
Parametros :
x = 0
y = 4
d = 0.1
n = 10
diff_eq = y_prime
func = y
"""

euler_method(0, 4, 0.1, 10, y_prime, y)

x_1 = 0.1000, y_1 = 3.2000;  error = 0.0749
x_2 = 0.2000, y_2 = 2.5600;  error = 0.1213
x_3 = 0.3000, y_3 = 2.0480;  error = 0.1472
x_4 = 0.4000, y_4 = 1.6384;  error = 0.1589
x_5 = 0.5000, y_5 = 1.3107;  error = 0.1608
x_6 = 0.6000, y_6 = 1.0486;  error = 0.1562
x_7 = 0.7000, y_7 = 0.8389;  error = 0.1475
x_8 = 0.8000, y_8 = 0.6711;  error = 0.1365
x_9 = 0.9000, y_9 = 0.5369;  error = 0.1243
x_10 = 1.0000, y_10 = 0.4295;  error = 0.1118


(0.4294967296000001, 0.1118444033464508)

### Solucion:

Utiliza esta tabla como guia para validar que tu funcion este correcta,
y reemplaza los valores que faltan con la salida de tu funcion.

| x     | 0 | 0.1         | 0.2    | 0.4    | 0.6    | 0.8    | 1            |
|-------|---|-------------|--------|--------|--------|--------|--------------|
| y     | 4 | 3.2749      | 2.6813 | 1.7973 | 1.2048 | 0.8076 | 0.5413       |
| $y_n$ | 4 | **3.200**       | 2.56   | 1.6384 | 1.0486 | 0.6711 | **0.4295**       |
| error | 0 | **0.0749**   | 0.1213 | 0.1589 | 0.1562 | 0.1365 | **0.1118** |

### 4.1 Probando la implementacion con una aplicación real de data science

Podemos utilizar ecuaciones diferenciales  como modelos matemáticos de fenónemos reales, por ejemplo si llegas a trabajar como data scientist para una farmaceutica  puede que interese analizar la desintegración radioactiva de cierta sustancia y como esto cambia en el tiempo
#### Desintegracion Radiactiva

$A$ es la cantidad de sustancia en el tiempo $t$, la derivada $\frac{\partial A}{\partial t}$ es rapidez a la que se desintegran los nucleos de una sustancia.

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial A}{\partial t} =  -0.56A \\
\end{equation*}
Donde $A(0) = 1$ vamos a un tamaño de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ A(t) = e^{-0.56t} \\
\end{equation*}

**Nota** En este caso sencillo, poseemos la solución exacta a la ED por lo cual no se necesita una solución aproximada , pero en aplicaciones y proyectos reales muchas veces la solución exacta no se posee, o es muy difícil de calcular y recurrimos a soluciones aproximadas por computadora , para este ejercicio el saber la solución exacta nos sirve para medir la exactitud de nuestra solución aproximada.

In [91]:
### Implementando las funciones para el ejemplo de desintegración radiactiva
def A_prime(t, A):
    return -0.56 * A

def A(t):
    return math.exp(-0.56*t)

In [92]:
"""
Llamamos a la función para aplicar el método de Euler

euler_method(0,4,0.1,10, y_prime,y)

Parámetros :
x = 0
y = 4
h = 0.1
n = 10
diff_eq = y_prime
func = y
"""

euler_method(0, 1, 0.1, 30, A_prime, A)

x_1 = 0.1000, y_1 = 0.9440;  error = 0.0015
x_2 = 0.2000, y_2 = 0.8911;  error = 0.0029
x_3 = 0.3000, y_3 = 0.8412;  error = 0.0041
x_4 = 0.4000, y_4 = 0.7941;  error = 0.0052
x_5 = 0.5000, y_5 = 0.7497;  error = 0.0061
x_6 = 0.6000, y_6 = 0.7077;  error = 0.0070
x_7 = 0.7000, y_7 = 0.6680;  error = 0.0077
x_8 = 0.8000, y_8 = 0.6306;  error = 0.0083
x_9 = 0.9000, y_9 = 0.5953;  error = 0.0088
x_10 = 1.0000, y_10 = 0.5620;  error = 0.0092
x_11 = 1.1000, y_11 = 0.5305;  error = 0.0096
x_12 = 1.2000, y_12 = 0.5008;  error = 0.0099
x_13 = 1.3000, y_13 = 0.4728;  error = 0.0101
x_14 = 1.4000, y_14 = 0.4463;  error = 0.0103
x_15 = 1.5000, y_15 = 0.4213;  error = 0.0104
x_16 = 1.6000, y_16 = 0.3977;  error = 0.0105
x_17 = 1.7000, y_17 = 0.3754;  error = 0.0105
x_18 = 1.8000, y_18 = 0.3544;  error = 0.0105
x_19 = 1.9000, y_19 = 0.3346;  error = 0.0105
x_20 = 2.0000, y_20 = 0.3158;  error = 0.0105
x_21 = 2.1000, y_21 = 0.2981;  error = 0.0104
x_22 = 2.2000, y_22 = 0.2814;  error = 0.0103
x_23 =

(0.17748425378021818, 0.00888972225919163)

Como vemos, conforme transcurre el tiempo, la sustancia radiactiva decae poco a poco. Se esperaría que en un tiempo muy largo ($t \rightarrow \infty$), la cantidad de sustancia $A \rightarrow 0$

## 5 - Rotacion de listas hacia la izquierda

Una rotacion a la izquierda en una lista mueve cada elemento de la lista una vez hacia la izquierda. Por ejemplo, si hacemos $1$ rotacion en en el arreglo $[1,2,3,4,5]$, el arreglo resultante seria $[2,3,4,5,1]$

#### Descipcion de la funcion
Dado un arreglo $a$ de $n$ integers y un numero $d$ de rotaciones, realizar $d$ rotaciones hacia la izquierda. La funcion debe retorner el arreglo re-ordenado.

**rotIzquierda** tiene los siguientes parametros:

* $a$ Un arreglo de numeros entreros.
* $d$ numero de rotaciones.

In [93]:
def rotIzquierda(a, d):
    # tu codigo aqui
    # Quitar el número de rotaciones en exceso del largo de la lista (por ejemplo: si la lista es [1,2,3,4,5], 
    #    tiene 5 elementos, y daría lo mismo rotar 7 veces hacia la izquierda, que 2 veces (7 % 5 = 2), pues 
    #    al rotar 5 veces sobre esta lista, esta regresa a su estado original
    d = d % len(a);
    
    # Devolver la lista
    return a[d:] + a[:d]

In [94]:
rotIzquierda([1, 2, 3, 4, 5], 4)
# Resultado: [5, 1, 2, 3, 4]

[5, 1, 2, 3, 4]

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

#### TIP:
Utiliza **slice** en el array para optimizar la funcion

In [95]:
validaciones.rotIzquierda_function(rotIzquierda)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
